### PR TITLE
Main prolog

### PR DESCRIPTION
**Pull request by ghislainfourny**
This makes MainModule and Prolog nodes, not expressions.

Also the TranslationVisitor is now refactored to cleanly return one node every time, which removes the intermediate result holders. This more closely reflects the visitor pattern.

### CODE REVIEW

File: src/main/java/org/rumbledb/expressions/module/MainModule.java [hunk 1]
```diff
@@ -22,22 +22,26 @@
...
     private final Prolog prolog;
     private final Expression expression;
 
-    public MainModule(Prolog prolog, Expression commaExpression, ExceptionMetadata metadata) {
+    public MainModule(Prolog prolog, Expression expression, ExceptionMetadata metadata) {
         super(metadata);
         this.prolog = prolog;
-        this.expression = commaExpression;
+        if (expression == null) {
+            throw new OurBadException("The main prolog must have a non-null expression");
```
Comment (CanBerker) [inline src/main/java/org/rumbledb/expressions/module/MainModule.java]: Is 'main prolog' intentional here?

File: src/main/java/org/rumbledb/expressions/postfix/PostfixExpression.java [hunk 2]
```diff
@@ -31,6 +32,9 @@
     public PostfixExpression(Expression mainExpression, ExceptionMetadata metadata) {
         super(metadata);
         this.mainExpression = mainExpression;
+        if (this.mainExpression == null) {
+            throw new OurBadException("Main expression cannot be null in a postfix expression.");
+        }
```
Comment (CanBerker) [inline src/main/java/org/rumbledb/expressions/postfix/PostfixExpression.java]: Very nice

File: src/test/java/iq/RuntimeTests.java [hunk 3]
```diff
@@ -94,7 +94,7 @@ public static void setupSparkSession() {
     public void testRuntimeIterators() throws Throwable {
         System.err.println(AnnotationsTestsBase.counter++ + " : " + this._testFile);
         TranslationVisitor visitor = new TranslationVisitor();
-        testAnnotations(this._testFile.getAbsolutePath(), visitor);
+        testAnnotations(this._testFile.getAbsolutePath());
```
Comment (CanBerker) [inline src/test/java/iq/RuntimeTests.java]: I just noticed that I missed a variable name with underscore here. Could you remove the underscore from the name?

File: src/test/java/iq/base/AnnotationsTestsBase.java [hunk 4]
```diff
@@ -95,11 +90,11 @@ public void initializeTests(File dir) {
             );
             if (this.currentAnnotation.shouldParse()) {
                 Assert.fail("Program did not parse when expected to.\nError output: " + errorOutput + "\n");
-                return context;
+                return null;
             } else {
                 System.out.println(errorOutput);
                 Assert.assertTrue(true);
```
Comment (CanBerker) [inline src/test/java/iq/base/AnnotationsTestsBase.java]: Does this line do anything ?

File: src/test/java/utils/AstSerialization.java [hunk 5]
```diff
@@ -21,6 +21,7 @@
 package utils;
 
 import org.rumbledb.compiler.TranslationVisitor;
```
Comment (CanBerker) [inline src/test/java/utils/AstSerialization.java]: This import can be removed

File: src/main/java/org/rumbledb/compiler/TranslationVisitor.java [hunk 6]
```diff
@@ -198,91 +188,79 @@ public Void visitFunctionDecl(JsoniqParser.FunctionDeclContext ctx) {
...
-        return null;
     }
     // endregion
 
     // region expr
     @Override
-    public Void visitExpr(JsoniqParser.ExprContext ctx) {
+    public Node visitExpr(JsoniqParser.ExprContext ctx) {
         List<Expression> expressions = new ArrayList<>();
         Expression expression;
```
Comment (CanBerker) [inline src/main/java/org/rumbledb/compiler/TranslationVisitor.java]: Unused variable can be removed

File: src/main/java/org/rumbledb/compiler/TranslationVisitor.java [hunk 7]
```diff
@@ -198,91 +188,79 @@ public Void visitFunctionDecl(JsoniqParser.FunctionDeclContext ctx) {
...
+            return this.visitFlowrExpr((JsoniqParser.FlowrExprContext) content);
+        if (content instanceof JsoniqParser.IfExprContext)
+            return this.visitIfExpr((JsoniqParser.IfExprContext) content);
+        if (content instanceof JsoniqParser.QuantifiedExprContext)
+            return this.visitQuantifiedExpr((JsoniqParser.QuantifiedExprContext) content);
+        if (content instanceof JsoniqParser.SwitchExprContext)
+            return this.visitSwitchExpr((JsoniqParser.SwitchExprContext) content);
+        if (content instanceof JsoniqParser.TypeSwitchExprContext)
+            return this.visitTypeSwitchExpr((JsoniqParser.TypeSwitchExprContext) content);
+        throw new OurBadException("Unrecognized ExprSingle.");
```
Comment (CanBerker) [inline src/main/java/org/rumbledb/compiler/TranslationVisitor.java]: Beautiful

File: src/main/java/org/rumbledb/compiler/TranslationVisitor.java [hunk 8]
```diff
@@ -458,265 +415,238 @@ public Void visitGroupByVar(JsoniqParser.GroupByVarContext ctx) {
...
-        this.visitAndExpr(ctx.main_expr);
-        if (!(ctx.rhs == null) && !ctx.rhs.isEmpty()) {
-            Expression mainExpression = this.currentExpression;
-            for (JsoniqParser.AndExprContext child : ctx.rhs) {
-                this.visitAndExpr(child);
-                rhs.add(this.currentExpression);
-            }
-            this.currentExpression = new OrExpression(mainExpression, rhs, createMetadataFromContext(ctx));
+        Expression mainExpression = (Expression) this.visitAndExpr(ctx.main_expr);
+        if (ctx.rhs == null || ctx.rhs.isEmpty()) {
```
Comment (CanBerker) [inline src/main/java/org/rumbledb/compiler/TranslationVisitor.java]: Much cleaner !

File: src/main/java/org/rumbledb/compiler/TranslationVisitor.java [hunk 9]
```diff
@@ -735,91 +665,81 @@ public Void visitPostFixExpr(JsoniqParser.PostFixExprContext ctx) {
...
         } else if (ctx.pe != null) {
-            this.visitParenthesizedExpr(ctx.pe);
+            return this.visitParenthesizedExpr(ctx.pe);
         } else if (ctx.vr != null) {
-            this.visitVarRef(ctx.vr);
+            return this.visitVarRef(ctx.vr);
         } else if (ctx.ci != null) {
-            this.visitContextItemExpr(ctx.ci);
+            return this.visitContextItemExpr(ctx.ci);
         } else if (ctx.tkw != null) {
```
Comment (CanBerker) [inline src/main/java/org/rumbledb/compiler/TranslationVisitor.java]: Shall we refactor these 'else if's to 'if's?

File: src/main/java/org/rumbledb/compiler/TranslationVisitor.java [hunk 10]
```diff
@@ -735,91 +665,81 @@ public Void visitPostFixExpr(JsoniqParser.PostFixExprContext ctx) {
...
         } else if (child instanceof TerminalNode) {
-            this.currentExpression = ValueTypeHandler.getValueType(child.getText(), createMetadataFromContext(ctx));
+            return ValueTypeHandler.getValueType(child.getText(), createMetadataFromContext(ctx));
         } else if (child instanceof JsoniqParser.ContextItemExprContext) {
-            this.visitContextItemExpr((JsoniqParser.ContextItemExprContext) child);
+            return this.visitContextItemExpr((JsoniqParser.ContextItemExprContext) child);
         } else if (child instanceof JsoniqParser.FunctionCallContext) {
-            this.visitFunctionCall((JsoniqParser.FunctionCallContext) child);
+            return this.visitFunctionCall((JsoniqParser.FunctionCallContext) child);
         } else if (child instanceof JsoniqParser.FunctionItemExprContext) {
```
Comment (CanBerker) [inline src/main/java/org/rumbledb/compiler/TranslationVisitor.java]: Same comment with else ifs

File: src/main/java/org/rumbledb/compiler/TranslationVisitor.java [hunk 11]
```diff
@@ -831,69 +751,57 @@ public Void visitObjectConstructor(JsoniqParser.ObjectConstructorContext ctx) {
...
-        return null;
     }
 
     @Override
-    public Void visitArrayConstructor(JsoniqParser.ArrayConstructorContext ctx) {
+    public Node visitArrayConstructor(JsoniqParser.ArrayConstructorContext ctx) {
         if (ctx.expr() == null)
-            this.currentExpression = new ArrayConstructorExpression(createMetadataFromContext(ctx));
+            return new ArrayConstructorExpression(createMetadataFromContext(ctx));
         else {
```
Comment (CanBerker) [inline src/main/java/org/rumbledb/compiler/TranslationVisitor.java]: Similarly, else can be removed

File: src/main/java/org/rumbledb/compiler/TranslationVisitor.java [hunk 12]
```diff
@@ -831,69 +751,57 @@ public Void visitObjectConstructor(JsoniqParser.ObjectConstructorContext ctx) {
...
-
     }
 
     @Override
-    public Void visitParenthesizedExpr(JsoniqParser.ParenthesizedExprContext ctx) {
+    public Node visitParenthesizedExpr(JsoniqParser.ParenthesizedExprContext ctx) {
         if (ctx.expr() == null)
-            this.currentExpression = new CommaExpression(createMetadataFromContext(ctx));
+            return new CommaExpression(createMetadataFromContext(ctx));
         else {
```
Comment (CanBerker) [inline src/main/java/org/rumbledb/compiler/TranslationVisitor.java]: Else can be removed

File: src/main/java/org/rumbledb/compiler/TranslationVisitor.java [hunk 13]
```diff
@@ -831,69 +751,57 @@ public Void visitObjectConstructor(JsoniqParser.ObjectConstructorContext ctx) {
...
-    public Void visitContextItemExpr(JsoniqParser.ContextItemExprContext ctx) {
-        this.currentExpression = new ContextItemExpression(createMetadataFromContext(ctx));
-        return null;
+    public Node visitContextItemExpr(JsoniqParser.ContextItemExprContext ctx) {
+        return new ContextItemExpression(createMetadataFromContext(ctx));
     }
 
     @Override
-    public Void visitSequenceType(JsoniqParser.SequenceTypeContext ctx) {
+    public Node visitSequenceType(JsoniqParser.SequenceTypeContext ctx) {
```
Comment (CanBerker) [inline src/main/java/org/rumbledb/compiler/TranslationVisitor.java]: I have noticed that SequenceType is currently not a Node. Is this due to the XQuery spec?

If this was a node as well we could get rid of the last field in the TranslationVisitor.

File: src/main/java/org/rumbledb/compiler/TranslationVisitor.java [hunk 14]
```diff
@@ -920,76 +828,69 @@ else if (ctx.plus.size() > 0)
...
-            else
-                this.currentExpression = new FlworVarSingleType(item, createMetadataFromContext(ctx));
-        }
-        return null;
+            return new FlworVarSingleType(createMetadataFromContext(ctx));
+
+        AtomicTypes item = FlworVarSingleType.getAtomicType(ctx.item.getText());
+        if (ctx.question.size() > 0)
+            return new FlworVarSingleType(item, true, createMetadataFromContext(ctx));
+        else
```
Comment (CanBerker) [inline src/main/java/org/rumbledb/compiler/TranslationVisitor.java]: else can be removed

File: src/main/java/org/rumbledb/compiler/TranslationVisitor.java [hunk 15]
```diff
@@ -920,76 +828,69 @@ else if (ctx.plus.size() > 0)
...
         return arguments;
     }
 
     @Override
-    public Void visitArgument(JsoniqParser.ArgumentContext ctx) {
+    public Node visitArgument(JsoniqParser.ArgumentContext ctx) {
         if (ctx.exprSingle() != null) {
-            this.visitExprSingle(ctx.exprSingle());
+            return this.visitExprSingle(ctx.exprSingle());
         } else {
```
Comment (CanBerker) [inline src/main/java/org/rumbledb/compiler/TranslationVisitor.java]: else can be removed

File: src/main/java/org/rumbledb/compiler/TranslationVisitor.java [hunk 16]
```diff
@@ -920,76 +828,69 @@ else if (ctx.plus.size() > 0)
...
     }
 
     @Override
-    public Void visitFunctionItemExpr(JsoniqParser.FunctionItemExprContext ctx) {
+    public Node visitFunctionItemExpr(JsoniqParser.FunctionItemExprContext ctx) {
         ParseTree child = ctx.children.get(0);
         if (child instanceof JsoniqParser.NamedFunctionRefContext) {
-            this.visitNamedFunctionRef((JsoniqParser.NamedFunctionRefContext) child);
+            return this.visitNamedFunctionRef((JsoniqParser.NamedFunctionRefContext) child);
         } else if (child instanceof JsoniqParser.InlineFunctionExprContext) {
```
Comment (CanBerker) [inline src/main/java/org/rumbledb/compiler/TranslationVisitor.java]: else if can be an if

File: src/main/java/org/rumbledb/expressions/module/MainModule.java (hunk 1 above)
Comment (ghislainfourny) [reply src/main/java/org/rumbledb/expressions/module/MainModule.java]: Good catch.

File: src/test/java/iq/RuntimeTests.java (hunk 3 above)
Comment (ghislainfourny) [reply src/test/java/iq/RuntimeTests.java]: Done

File: src/test/java/iq/base/AnnotationsTestsBase.java (hunk 4 above)
Comment (ghislainfourny) [reply src/test/java/iq/base/AnnotationsTestsBase.java]: good catch!

File: src/test/java/utils/AstSerialization.java (hunk 5 above)
Comment (ghislainfourny) [reply src/test/java/utils/AstSerialization.java]: Done

File: src/main/java/org/rumbledb/compiler/TranslationVisitor.java (hunk 6 above)
Comment (ghislainfourny) [reply src/main/java/org/rumbledb/compiler/TranslationVisitor.java]: Done

File: src/main/java/org/rumbledb/compiler/TranslationVisitor.java (hunk 9 above)
Comment (ghislainfourny) [reply src/main/java/org/rumbledb/compiler/TranslationVisitor.java]: Nice, done.

File: src/main/java/org/rumbledb/compiler/TranslationVisitor.java (hunk 11 above)
Comment (ghislainfourny) [reply src/main/java/org/rumbledb/compiler/TranslationVisitor.java]: Done

File: src/main/java/org/rumbledb/compiler/TranslationVisitor.java (hunk 12 above)
Comment (ghislainfourny) [reply src/main/java/org/rumbledb/compiler/TranslationVisitor.java]: Done

File: src/main/java/org/rumbledb/compiler/TranslationVisitor.java (hunk 13 above)
Comment (ghislainfourny) [reply src/main/java/org/rumbledb/compiler/TranslationVisitor.java]: It is not (it is only a property of some expressions). However, I nevertheless was able to get rid of the field, via a util method.

File: src/main/java/org/rumbledb/compiler/TranslationVisitor.java (hunk 14 above)
Comment (ghislainfourny) [reply src/main/java/org/rumbledb/compiler/TranslationVisitor.java]: Yes, adding {} automatically will be best once it is merge. Thanks!

File: src/main/java/org/rumbledb/compiler/TranslationVisitor.java (hunk 16 above)
Comment (ghislainfourny) [reply src/main/java/org/rumbledb/compiler/TranslationVisitor.java]: Done

File: src/main/java/org/rumbledb/compiler/TranslationVisitor.java (hunk 15 above)
Comment (ghislainfourny) [reply src/main/java/org/rumbledb/compiler/TranslationVisitor.java]: Done

File: src/main/java/org/rumbledb/compiler/TranslationVisitor.java (hunk 13 above)
Comment (CanBerker) [reply src/main/java/org/rumbledb/compiler/TranslationVisitor.java]: Very nice

**Comment by ghislainfourny** [comment]
Thank you!

**Pull request merged**